xapian-maintainer-tools/profiling/profiler.c

```diff
@@ -57,15 +72,17 @@ typedef int (*real_open_t)(const char *, int, ...);
 int open(const char *pathname, int flags, ...)
 {
     int fd;
+    if (!cached_function_ptrs[CALL_OPEN]) {
+	cached_function_ptrs[CALL_OPEN] = dlsym(RTLD_NEXT, "open");
+    }
```

Comment:
I think this would be cleaner if you just use a function-scope static variable for each cached pointer, i.e. something like:

```
static real_open_t real_open = NULL;
if (!real_open) {
    real_open = (real_open_t)dlsym(RTLD_NEXT, "open");
}
```

(If this was C++, you could just write: `static real_open_t real_open = (real_open_t)dlsym(RTLD_NEXT, "open");` as C++ allows arbitrary expressions for initialising a static variable - the expression is evaluation the first time (and only the first time) that control flow reaches the statement - see: https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables.  But prob ably it's better to stick to C for this code.)

With this approach the function pointer has the appropriate type so there's no need for the visually cumbersome cast at the places where it is called - you can just write `fd = real_open(pathname, flags);` etc.

Reply:
Changed accordingly.